### ISSUE TITLE
Please reconsider; do not set --one-file-system in config.pl as default

### ISSUE THREAD

**Issue by ghost** [backuppc/backuppc#437]
In relation to https://sourceforge.net/p/backuppc/mailman/message/37326354/ , commits 02c33dab54f3538e82d8c61376eb5c49f6b8dbfc and d9a448960a9ae31edcec912308e52d471977b848 , please do not use --one-file-system as the default rsync argument in config.pl . 

If i understand the workings of this option correctly, this results in silently excluding whole filesystems from backup in quite common setups with i.e. multiple LVM logical volumes or ZFS filesystems. For a backup-solution, having too much backup data  (oops, forgot to exclude a mount or those /proc-thingies just give errors, let's exclude them by default) should always be preferred over having too little (oops, this hasn't been backed up since 2017 because I forgot to add it).

**Comment by hamster65** [comment]
I agree. A discovered by accident that my new BackupPC server had not backed up my laptop for several month because after I switched on ecryptfs for the home directory.
Luckily it was covered by the old BackupPC-Server, which was still running and had, for some reason, that option disabled.

**Comment by wolfgangkarall** [comment]
I'm not sure I understand the rationale of the change, as the comments in both the old v3 and new v4 config file near RsyncShareName should be enough to give an admin the hint on how to use it in case it makes sense to their scenario, whereas having this new default means you don't get complete backups on systems with anything but the most basic single-filesystem layout even on new setups.

That this all happens silently makes this even worse; if there would be some way to raise a bug's priority I would in this case use at least 'important', maybe even 'grave' (according to Debian's severity levels), as there might be data loss in the 'no backed up data to restore from'-case.

I've been hit by this on Debian [1] where backups for 90+ servers (all with multiple file systems, as we're the kind of old school team where this is still done) were not done properly after upgrading to v4. Luckily this was spotted without actually wanting to restore data.

[1] where the upgrade routine is not doing anything like https://github.com/backuppc/backuppc/commit/d9a448960a9ae31edcec912308e52d471977b848 on upgrade to at least have the option gone from the .dpkg-dist file, which we take as base for re-applying manual changes after a major upgrade like this.

**Comment by haarp** [comment]
Agreed, this setting seems to cause a lot of confusion and new issues. See e.g. #447 or #459, which are likely this.

**Comment by Barnoid-** [comment]
I strongly second this request. I have just been bitten by this after an upgrade from 3.3 to 4.4. I didn't notice that the encrypted user home was not backed up anymore after the upgrade because all backups report no errors and one can see the (old) files in the most backup.

**Comment by whuut** [comment]
I support this as well. This configuration change should have at least been mentioned in the "What's new in v4" section of the documentation.

I think it was a common approach in v3 to use RsyncShareName / and just maintain a shared list of exludes (like /proc) across different servers. This way you don't have to worry about the mount points of each server and always have everything backed up. For example, /var is often a separate mount point and contains most user data. This approach is especially handy if you are automating backups for a fleet of servers.

Fortunately, I discovered this accidentally before we needed backups, but that could have been a disaster.

**Comment by TheCrockett** [comment]
I’d like to add my objections as well. 

Others have articulated why above

**Comment by AHVrentner** [comment]
I can't believe, that someone has such a (non-)brilliant idea to even think of and rather actually adding such an option nice and silent and quietly... 
F...K!!
I'm speaking too of  the rsync option  "--one-file-system".
**I just lost backups for good after I finally decided to upgrade to 4.4 from a 3.n.**

How can you do that without even making a huge warning: **YOU WILL LOOSE DATA WITH THIS CONFIG-OPTION ENABLED!**

Not amused at all. Can you tell?!
I mean. Does one need to have a backup of a backup of a backup, when one upgrades a simple version?!
I don't think so!! But introducing silently an option like this, someone does not understand the implication of such a thing at all.
Please take note: Mounted filesystems on different paths no matter they're encrypted or not, is standard procedure in a *nix environment! It's absolutely irresponsible to introduce that in this way!!! You better would have said: Erase your old backups and start from fresh! After all we will erase them anyway. (Cynicism off!)

Sorry for being so critically. Just really venting my frustration about it.
Some quality managemnt was obviously in vacation. Must have been an intern bundlingn that package together.

**Comment by adamgoryachev** [comment]
Quality Management/Change Control goes both ways, make sure you review and understand the changes you are implementing, before proceeding to do so. After making a big change to your backup system, verify that it is still working as intended. 
Consider, --one-file-system is not a bad option, in itself. Sure, it might be normal to use multiple mount points, but adding those to your backup system is just a standard process of adding it to your server (ie, /etc/fstab or equivalent). There will always be differing opinions on what should be a default setting, hence, make sure you carefully review the defaults you use to ensure they still meet your requirements in your environment.
Most likely, this will just include a bunch of people complaining about the current default setting, for everyone else, they are happy with the default. So even if there are hundreds of complaints, it does not indicate a bad default by itself.

Just my 0.02c worth

**Comment by AHVrentner** [comment]
> Quality Management/Change Control goes both ways, make sure you review and understand the changes you are implementing, before proceeding to do so. After making a big change to your backup system, verify that it is still working as intended. Consider, --one-file-system is not a bad option, in itself. Sure, it might be normal to use multiple mount points, but adding those to your backup system is just a standard process of adding it to your server (ie, /etc/fstab or equivalent). There will always be differing opinions on what should be a default setting, hence, make sure you carefully review the defaults you use to ensure they still meet your requirements in your environment. Most likely, this will just include a bunch of people complaining about the current default setting, for everyone else, they are happy with the default. So even if there are hundreds of complaints, it does not indicate a bad default by itself.
> 
> Just my 0.02c worth

Well sorry, Adam, your 0.02c worth is actually worth 0.000c (i.e. nothing).  ;-)
Why?
Nowadays, systems are updateing themselves all automatically. After all, everyone says one needs to do that because of security and .... what not....
BackupPC does not make an exception here. And before you know it  you get a nice and elegant "new feature" introduced that you start to realize then when stuff is missing... 
Afte rall, one has also other things to do then just read changelogs of all the applications that get updated and upgraded almost by the hour ...
Do I reallyl need to spell it out? 
I don't think so.

So, I entirely disagree with you, and agree with my former respondent Barnoid!

**Comment by TheCrockett** [comment]
I think it comes down to a fundamental question. 

"What is the intention of the software", which is pretty easy. "Back up your stuff". Okay cool, so if we can all agree on that; then the subsequent implementation question is:

"What method are we controlling the 'what gets backed up'." with?  Now, having used BackupPC for at least since 2008, the assumption has always been "Your rsync 'share name' (in the case of rsyncd), controls the configuration of the 'what'. 

Ex:

[root]
path=/

or

[just-my-pictures]
path=/mnt/pictures

If we say that the logic of 'what' is fundamentally controlled by rsyncd.conf; then I put forth that the configuration (i.e. --one-file-system) should also respect the spirit of the share-level configuration. 

in short - we should not use one-file-system because that logic is controlled at the share name level.

**Comment by mpounsett** [comment]
I am astonished that this has been under discussion for so long and it's still the default.

Adding this into the default configuration without at least a big bright shiny flashing red alarm at the top of the release notes is actively hostile, but it's not even mentioned in the 'changes in 4.0' section of the docs.

**Comment by GWHAYWOOD** [comment]
Closing - see d9a448960a9ae31edcec912308e52d471977b848.

**Comment by TheCrockett** [comment]
This should not be closed. 

A hidden comment in the source code is not sufficent here for people who depend on this as a backup

**Comment by wolfgangkarall** [comment]
@GWHAYWOOD I'm sorry, but how can pointing to a 8 year old commit (that was referenced already in a comment above as not being a solution) be considered a reason to close this issue?

**Comment by mpounsett** [comment]
This issue is not completed.

**Comment by TheCrockett** [comment]
@mpounsett @wolfgangkarall 

It might be time to fork BackupPC if this is the kind of care and feedback that we’re going to get for something this fundamental to our ecosystems

**Comment by craigbarratt** [comment]
I realize adding `--one-file-system` as the default in 4.x has been frustrating.  The reason I did that was that sparse and device  files on linux systems can be extremely large, which causes backups to run seemingly indefinitely, and the pool to get filled with very large files (if it even gets that far).  Try backing up `/proc` or `/dev` (or `/var/log/wtmp` on certain linuxes) and see how well things go for you.

So it's a tradeoff between having an incorrectly configured host backing up less than expected, or failing badly and opaquely in a way that could eventually cause all backups to fail.

The solution is taking a few seconds to configure the host settings correctly, and of course taking a few seconds to confirm a successful backupc captures what you expect.

We'd certainly consider PRs that remove `--one-file-system` as the default, but they would have to reliably avoid large sparse files (and device and proc files).  It's not sufficient to just exclude `/proc` and `/dev` since large sparse or special files can lurk anywhere.

@TheCrockett writes:

> It might be time to fork BackupPC if this is the kind of care and feedback that we’re going to get for something this fundamental to our ecosystems

You are most welcome to fork BackupPC.  It's OSS after all.  But you might be underestimating the amount of time required to support and maintain such a system.  Perhaps you'd be happier (and I'd be happier) if you switch to a different backup solution?

**Comment by Barnoid-** [comment]
@craigbarratt While I understand the issues with backing up special filesystems like /dev or /proc, I respectfully disagree with the decision to simply exclude problematic directories by using the --one-file-system directive.

While this approach may avoid certain edge cases, it introduces more serious concerns:

1. It doesn't address the issue of large sparse files. And even so - why shouldn’t those be backed up?
2. It has already caused (and is likely to continue causing) confusion and data loss for BackupPC users, many of whom are experienced Linux administrators.

Personally, I’d much rather have a backup process that fails loudly (and gets stuck) than one that silently skips parts of the directory tree - especially as a default behavior. If BackupPC hangs, I can investigate, adjust my include/exclude list, and re-run the backup. But if it silently omits parts of my data, I may not even realize something is missing.

In fact, as I noted in a comment years ago, I once moved portions of my /home directory to a separate partition due to disk space limitations. Those files are now silently ignored because of the --one-file-system default. They still show up in the BackupPC interface under older backup IDs, making this behavior especially subtle and hard to detect unless I do full directory restores and comparisons.

tl;dr: If certain directories (like /proc, /dev, /tmp, etc.) are problematic, exclude them explicitly by default. If large files are a concern, recommend users use the --max-size option. But please reconsider the use of --one-file-system as a default - it’s a misleading and potentially dangerous choice for a backup tool.

**Comment by ghost** [comment]
While I certainly respect above explanations when in favor of the current behavior, I certainly agree with the suggestion of not closing this issue. Regardless of the preference; I believe a very common configuration having multiple fs's (ZFS datasets, LVM LV's, et cetera) AND new fs's due to adding an extra LV now NOT being back-upped by default is a lot more common then issues due to unforeseen stuff like spares files. This will be an issue either way that needs a much more clear approach; i.e. make clear to the user what is the default, what is the expected result, et cetera. In it's current state, we're either surprised by a backup program that by default doesn't backup everything when requested  (including newly added filesystems) OR a backup program that when we fix/change the default due to this issue, suddenly causes issues due to i.e. spare files.

Please, for now at least reopen the issue for visibility.

P.S.: Thanks @craigbarratt for all your years of effort and your help in the transfer to @GWHAYWOOD. And @GWHAYWOOD thank you so much for taking this project on your shoulders.
